src/tools/rbd_mirror/image_map/Policy.cc

```diff
@@ -0,0 +1,239 @@
+}
+
+template <typename I>
+void Policy<I>::add_instances(const std::vector<std::string> &instances) {
+  dout(20) << dendl;
+
+  assert(m_map_lock.is_wlocked());
+
+  for (auto const &instance : instances) {
+    dout(10) << ": adding instance_id=" << instance << dendl;
+    m_map.emplace(instance, std::set<std::string>{});
+  }
+}
+
+template <typename I>
+void Policy<I>::remove_instances(const std::vector<std::string> &instances) {
+  dout(20) << dendl;
+
+  assert(m_map_lock.is_wlocked());
+
+  for (auto const &instance : instances) {
+    dout(10) << ": removing instance_id=" << instance << dendl;
+    if (!m_map[instance].empty()) {
+      dout(5) << ": instance_id=" << instance << " still has " << m_map[instance].size()
+              << " mapped image(s), continuing removal" << dendl;
+    }
+
+    m_map.erase(instance);
+  }
+}
+
+template <typename I>
+void Policy<I>::shuffle_add_instances(const std::vector<std::string> &instances,
+                                      std::map<std::string, std::pair<std::string, std::string> > *remap_ids) {
+  dout(20) << ": adding " << instances.size() << " instance(s)" << dendl;
+
+  assert(m_map_lock.is_wlocked());
+
+  add_instances(instances);
+  do_shuffle_add_instances(instances, remap_ids);
+
+  utime_t now = ceph_clock_now();
+  CephContext *cct = reinterpret_cast<CephContext *>(m_ioctx.cct());
+  int migration_throttle = cct->_conf->get_val<int64_t>("rbd_mirror_image_policy_migration_throttle");
+
+  dout(10) << ": migration_throttle=" << migration_throttle << " sec(s)" << dendl;
+
+  for (auto it = (*remap_ids).begin(); it != (*remap_ids).end();) {
+    const std::string &global_image_id = (*it).first;
+
+    if (!can_shuffle_image(now, m_shuffled_timestamp[global_image_id], migration_throttle)) {
+      dout(5) << ": not shuffling " << global_image_id << " (last_shuffle_time="
+              << m_shuffled_timestamp[global_image_id] << ")" << dendl;
+
+      it = (*remap_ids).erase(it);
+      continue;
+    }
+
+    remap(global_image_id, (*it).second.first, (*it).second.second, now);
+    ++it;
+  }
+}
+
+template <typename I>
+void Policy<I>::shuffle_remove_instances(const std::vector<std::string> &instances,
+                                         std::map<std::string, std::pair<std::string, std::string> > *remap_ids) {
+  dout(20) << ": removing " << instances.size() << " instance(s)" << dendl;
+
+  assert(m_map_lock.is_wlocked());
+
+  do_shuffle_remove_instances(instances, remap_ids);
+
+  utime_t now = ceph_clock_now();
+  for (auto const &kv : *remap_ids) {
+    remap(kv.first, kv.second.first, kv.second.second, now);
+  }
+
+  remove_instances(instances);
+}
+
+template <typename I>
+void Policy<I>::shuffle(const std::vector<std::string> &instances,
+                        std::map<std::string, std::pair<std::string, std::string> > *remap_ids) {
+  dout(20) << dendl;
+
+  // for now this doesn't support updating instances with mix of
```

Comment:
I'd say this should either be broken into ```add_instance``` and ```remove_instance``` methods or it should handle mixed add/remove cases. The former would probably be easier (and cleaner). This would impact ```ImageMap```'s interface as well.